File: lib/bloom_cache.rb
require 'ffi'

module BloomCacheFFI
  extend FFI::Library    
  ffi_lib $LOAD_PATH.map{|p| p + '/bloom_cache/bloom_cache.so'}
  attach_function :create_bloom_cache, [:int, :int, :int], :pointer
  attach_function :print_bloom_cache, [:pointer], :void
  attach_function :add_value_to_bucket, [:pointer, :int, :int, :int], :void
  attach_function :add_element, [:pointer, :pointer, :int, :int], :void
  attach_function :retrieve_elements, [:pointer, :pointer], :pointer
  attach_function :add_value, [:pointer, :pointer, :int, :int, :int], :void
  attach_function :match_elements, [:pointer, :pointer, :int], :pointer
  attach_function :delete_bloom_cache, [:pointer], :void
  attach_function :delete_result_set, [:pointer], :void
  attach_function :sort_buckets, [:pointer], :void
  attach_function :clear_cache, [:pointer], :void
  attach_function :merge_results, [:pointer, :pointer, :int, :pointer], :void
  attach_function :create_result_set, [:int], :pointer
  attach_function :add_results, [:pointer, :pointer, :int], :void
end

class ResultSet < FFI::ManagedStruct
  layout :data, :pointer,
    :count, :int,
    :size, :int,
    :size_step, :int

  [:map, :each].each do |meth|
    define_method meth do |*args, &block|
      to_a.send(meth, *args, &block)
    end
  end

  def push(*vals)
    concat(vals.flatten)
  end

  def concat(other)
    raise ArgumentError('Target must convert to an array') if !other.respond_to?(:to_a)
    arr = other.to_a
    p = FFI::MemoryPointer.new(:int32, arr.length)
    p.write_array_of_int32(arr)
    BloomCacheFFI.add_results(pointer, p, arr.length)
  end

  def length
    self[:count]
  end

  def count
    self[:count]
  end

  def to_a
    self[:data].get_array_of_int(0, self[:count])
  end

  def self.create(initial_size)
    ResultSet.new(BloomCacheFFI.create_result_set(initial_size))
  end

  def self.release(ptr)
    BloomCacheFFI.delete_result_set(ptr)
  end

  def clear
    self[:count] = 0
  end

  def inspect
    to_a.inspect
  end

  def self.merge_results(results, destination)
    raise ArgumentError('Destination must be a ResultSet') if !destination.is_a? ResultSet      
    res_count_array = FFI::MemoryPointer.new(:int32, results.count)
    res_count_array.write_array_of_int32(results.map(&:length))
    res_pointers = []
    results.each do |res|
      p = FFI::MemoryPointer.new(:int32, res.count)
      p.write_array_of_int32(res)
      res_pointers.push(p)
    end
    res_pointer_buf = FFI::MemoryPointer.new(:pointer, results.count)
    res_pointer_buf.write_array_of_pointer(res_pointers)
    BloomCacheFFI.merge_results(res_pointer_buf, res_count_array, results.length, destination)
  end

end

class BloomCache < FFI::ManagedStruct

  layout :bitwidth, :int,
    :intwidth, :int,
    :buckets, :pointer,
    :bucket_lengths, :pointer,
    :bucket_counts, :pointer,
    :bucket_length_step, :int,
    :max_bucket_count, :int,
    :hash_count, :int

  def buckets
    self[:buckets].get_array_of_pointer(0, self[:bitwidth]).map.with_index do |p, i|
      p.get_array_of_int(0, self[:bucket_counts].get_array_of_int(0, self[:bitwidth])[i])
    end
  end

  def clear
    BloomCacheFFI.clear_cache(pointer)
  end

  def sort_buckets
    BloomCacheFFI.sort_buckets(pointer)
  end

  def self.create(bitwidth, bucket_length_step, hash_count)
    BloomCache.new(BloomCacheFFI.create_bloom_cache(bitwidth, bucket_length_step, hash_count))
  end

  def print
    BloomCacheFFI.print_bloom_cache(pointer)
  end

  def add_value_to_bucket(bucket, value, sort = false)
    val = sort ? 1 : 0
    BloomCacheFFI.add_value_to_bucket(pointer, bucket, value, val)
  end

  def add_value(string, id, sort = false)
    buf = FFI::MemoryPointer.new(:uint8, string.bytes.length)
    buf.put_array_of_uint8(0, string.bytes)
    BloomCacheFFI.add_value(pointer, buf, string.bytes.length, id, sort ? 1 : 0)
  end

  def add_element(bitfield, value)        
    BloomCacheFFI.add_element(pointer, bitfield_to_buffer(bitfield), value)
  end

  def retrieve_elements(bitfield) 
    ResultSet.new(BloomCacheFFI.retrieve_elements(pointer, bitfield_to_buffer(bitfield)))
  end

  def match(string)
    buf = FFI::MemoryPointer.new(:uint8, string.bytes.length)
    buf.put_array_of_uint8(0, string.bytes)
    ResultSet.new(BloomCacheFFI.match_elements(pointer, buf, string.bytes.length))
  end

  def bitfield_to_buffer(bitfield)
    int_width = [(self[:bitwidth].to_i / 32), 1].max
    puts "int_width: #{int_width}"   
    buf = FFI::MemoryPointer.new(:uint32, int_width)
    arr = []
    int_width.times do |i|
      mask = (1 << 32) - 1
      mask <<= (i * 32)
      arr.push((bitfield & mask) >> (i * 32))
    end
    buf.put_array_of_uint32(0, arr)    
    buf
  end

  def self.release(ptr)
    BloomCacheFFI.delete_bloom_cache(ptr)
  end

end

File: ext/bloom_cache/bloom_cache.c
#include "bloom_cache.h"
#include "result_set.h"
#include "common.h"
#include "hashing.h"
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

void populate_result_set(BloomCache* b, ResultSet* rs, int* matching_buckets, int count);
void set_bit(uint32_t* bitfield, int position);
void add_element(BloomCache* b, uint32_t* bitfield, int value, int sort);

void Init_bloom_cache() {

}

void clear_cache(BloomCache* b) {
  int i;
  for (i = 0; i < b->bitwidth; ++i) {
    b->bucket_counts[i] = 0;
  }
}

BloomCache* create_bloom_cache(int bitwidth, int bucket_length_step, int hash_count) {
  BloomCache* res;
  int i;
  res = malloc(sizeof(BloomCache));  
  res->intwidth = ((bitwidth - 1) / 32) + 1;
  res->bitwidth = bitwidth;
  res->buckets = malloc(sizeof(int*) * res->bitwidth);
  res->bucket_lengths = malloc(sizeof(int) * res->bitwidth);
  res->bucket_counts = malloc(sizeof(int) * res->bitwidth);
  res->bucket_length_step = bucket_length_step;
  res->max_bucket_count = bucket_length_step;
  res->hash_count = hash_count;
  res->hash_service = create_hash_service(res->hash_count * 4);
  for (i = 0; i < bitwidth; ++i) {
    res->buckets[i] = malloc(sizeof(int) * bucket_length_step);
    res->bucket_lengths[i] = bucket_length_step;
    res->bucket_counts[i] = 0;
  }
  return res;
}

void delete_bloom_cache(BloomCache* b) {
  int i;
  delete_hash_service(b->hash_service);
  free(b->bucket_lengths);
  free(b->bucket_counts);
  for (i = 0; i < b->bitwidth; ++i) {
    free(b->buckets[i]);    
  }
  free(b->buckets);
  free(b);
}

void print_bloom_cache(BloomCache* b) {
  int i;
  printf("%d Buckets in total\n\n", b->bitwidth);
  for(i = 0; i < b->bitwidth; ++i) {
    printf("%d", b->bucket_counts[i]);
    /*printf("Bucket %d, length: %d, count: %d\n\n", i, b->bucket_lengths[i], b->bucket_counts[i]);    
    for (j = 0; j < b->bucket_counts[i]; ++j) {
      printf("%d\n", b->buckets[i][j]);
    }
    */
  }
  printf("\n");
}

void bitfield_from_string(BloomCache* b, const char* str, int data_length, uint32_t* bitfield) {
  int i, offset;
  unsigned char* res;

  res = malloc(sizeof(char) * b->hash_service->digest_size);
  memset(bitfield, 0, sizeof(uint32_t) * b->intwidth);
  if (data_length < 3) {
    for (offset = 0; offset < data_length; ++offset) {
      hash_data(b->hash_service, str + offset, 1, res);
      for (i = 0; i < b->hash_count; ++i) {
        set_bit(bitfield, *(unsigned int*)(res + (i * 4)) % b->bitwidth);
      }
    }
  }
  else {
    for (offset = 0; offset < data_length - 2; ++offset) {
      hash_data(b->hash_service, str + offset, 3, res);
      for (i = 0; i < b->hash_count; ++i) {
        set_bit(bitfield, *(unsigned int*)(res + (i * 4)) % b->bitwidth);
      }
    }
  }
    
  free(res);
}

void add_value(BloomCache* b, const char* data, int data_length, int id, int sort) {
  uint32_t* bitfield;
  
  bitfield = malloc(sizeof(uint32_t) * b->intwidth);
  bitfield_from_string(b, data, data_length, bitfield);
  add_element(b, bitfield, id, sort);
  free(bitfield);
}

void add_element(BloomCache* b, uint32_t* bitfield, int value, int sort) {
  int i;
  for(i = 0; i < b->bitwidth; ++i) {
    if (is_set(bitfield, i)) {
      add_value_to_bucket(b, i, value, sort);      
    }    
  }
}

void add_value_to_bucket(BloomCache* b, int bucket, int value, int sort) {
  if (b->bucket_counts[bucket] >= b->bucket_lengths[bucket]) {
    resize_bucket(b, bucket);
  }
  b->buckets[bucket][b->bucket_counts[bucket]] = value;
  ++b->bucket_counts[bucket];
  if (sort) {
    qsort(b->buckets[bucket], b->bucket_counts[bucket], sizeof(int), comparison);
  }
}

void sort_buckets(BloomCache* b) {
  int bucket = 0;
  for (bucket = 0; bucket < b->bitwidth; ++bucket) {
    qsort(b->buckets[bucket], b->bucket_counts[bucket], sizeof(int), comparison);
  }
}

void resize_bucket(BloomCache* b, int bucket) {
  int new_size = b->bucket_lengths[bucket] + b->bucket_length_step;
  b->bucket_lengths[bucket] = new_size;
  b->buckets[bucket] = realloc(b->buckets[bucket], sizeof(int) * new_size);
  if (new_size > b->max_bucket_count) {
    b->max_bucket_count = new_size;
  }
}

ResultSet* retrieve_elements(BloomCache* b, uint32_t* bitfield) {
  int* matching_buckets;
  int i, count;
  ResultSet* res;

  count = 0;  
  matching_buckets = malloc(sizeof(int*) * b->bitwidth);
  for (i = 0; i < b->bitwidth; ++i) {
    if (is_set(bitfield, i)) {
      matching_buckets[count] = i;
      ++count;
    }
  }
  res = create_result_set(b->max_bucket_count * 2);
  populate_result_set(b, res, matching_buckets, count);
  free(matching_buckets);
  return res;
}

ResultSet* match_elements(BloomCache* b, const char* value, int data_length) {
  uint32_t* bitfield;
  //int i;
  ResultSet* res;
  bitfield = malloc(sizeof(uint32_t) * b->intwidth);
  bitfield_from_string(b, value, data_length, bitfield);
  res = retrieve_elements(b, bitfield);
  free(bitfield);
  return res;
}

void populate_result_set(BloomCache* b, ResultSet* rs, int* matching_buckets, int count) {
  int i;
  int** data;
  int* counts;
  data = malloc(sizeof(int*) * count);
  counts = malloc(sizeof(int) * count);
  for (i = 0; i < count; ++i) {
    data[i] = b->buckets[matching_buckets[i]];
    counts[i] = b->bucket_counts[matching_buckets[i]];
  }
  merge_results(data, counts, count, rs);  
}

int is_set(uint32_t* bitfield, int position) {
  int field_part = position / 32;
  int field_offset = position % 32;
  return (bitfield[field_part] & (1 << field_offset)) >> field_offset;
}

void set_bit(uint32_t* bitfield, int position) {
  int field_part = position / 32;
  int field_offset = position % 32;
  bitfield[field_part] |= (1 << field_offset);
}